samples/nrf9160/serial_lte_modem/src/mqtt/th_mqtt_service.c

```diff
@@ -0,0 +1,633 @@
+ */
+static int fds_init(struct mqtt_client *c)
+{
+	if (c->transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		fds.fd = c->transport.tcp.sock;
+	} else {
+#if defined(CONFIG_MQTT_LIB_TLS)
+		fds.fd = c->transport.tls.sock;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	fds.events = POLLIN;
+
+	return 0;
+}
+
+static int init_mqtt_client(const u8_t *client_id, u8_t length)
+{
+	int err;
+
+	/* mqtt_client_init does not return */
+	mqtt_client_init(&client);
+
+#if defined(CONFIG_THIN_MQTT_USE_IP)
+	err = util_parse_host_by_ipv4(CONFIG_THIN_MQTT_BROKER_HOST_IP,
+		CONFIG_THIN_MQTT_BROKER_PORT, &broker);
+#else
+	err = util_parse_host_by_name(CONFIG_THIN_MQTT_BROKER_HOST_NAME,
+		CONFIG_THIN_MQTT_BROKER_PORT, &broker);
+#endif
+	if (err) {
+		return err;
+	}
+
+	/* MQTT client configuration */
+	client.broker = (struct sockaddr *)&broker;
+	client.evt_cb = mqtt_evt_handler;
+	client.client_id.utf8 = (u8_t *)client_id;
+	client.client_id.size = length;
+#ifdef CONFIG_THIN_MQTT_BROKER_USER_NAME
+	if (strlen(CONFIG_THIN_MQTT_BROKER_USER_NAME) > 0) {
+		client.user_name = k_malloc(sizeof(struct mqtt_utf8));
+		__ASSERT_NO_MSG(client.user_name != 0);
+		client.user_name->utf8 = k_malloc(64);
+		__ASSERT_NO_MSG(client.user_name->utf8 != 0);
+		memcpy(client.user_name->utf8,
+			(u8_t *)CONFIG_THIN_MQTT_BROKER_USER_NAME,
+			strlen(CONFIG_THIN_MQTT_BROKER_USER_NAME));
+		client.user_name->size =
+			strlen(CONFIG_THIN_MQTT_BROKER_USER_NAME);
+	} else {
+		client.user_name = NULL;
+	}
+#else
+	client.user_name = NULL;
+#endif
+#ifdef CONFIG_THIN_MQTT_BROKER_PASSWORD
+	if (strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD) > 0) {
+		client.password = k_malloc(sizeof(struct mqtt_utf8));
+		__ASSERT_NO_MSG(client.password != 0);
+		client.password->utf8 = k_malloc(64);
+		__ASSERT_NO_MSG(client.password->utf8 != 0);
+		memcpy(client.password->utf8,
+			(u8_t *)CONFIG_THIN_MQTT_BROKER_PASSWORD,
+			strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD));
+		client.password->size =
+			strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD);
+	} else {
+		client.password = NULL;
+	}
+#else
+	client.password = NULL;
+#endif
+	client.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client.rx_buf = rx_buffer;
+	client.rx_buf_size = sizeof(rx_buffer);
+	client.tx_buf = tx_buffer;
+	client.tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	client.transport.type = MQTT_TRANSPORT_NON_SECURE;
+	return 0;
+}
+
+static void uninit_mqtt_client(void)
+{
+	client.broker = NULL;
+	if (client.user_name) {
+		k_free(client.user_name->utf8);
+		k_free(client.user_name);
+	}
+	if (client.password) {
+		k_free(client.password->utf8);
+		k_free(client.password);
+	}
+	fds.fd = INVALID_FDS;
+	k_sem_reset(&mqtt_connected_sem);
+	mqtt_connected = false;
+}
+
+int do_mqtt_connect(const u8_t *client_id, u8_t length)
+{
+	int err;
+
+#if defined(CONFIG_THIN_MQTT_USE_IP)
+	LOG_INF("connect to %s:%d",
+		log_strdup(CONFIG_THIN_MQTT_BROKER_HOST_NAME),
```

Comment:
I suppose `CONFIG_THIN_MQTT_BROKER_HOST_IP` was meant to be used here?

Reply:
What a bug :)